.github/scripts/transform_dbt_test_results.py

```diff
@@ -1052,24 +1055,62 @@ def get_test_categories(
-                        and leg.{TAXYR_FIELD} = test_results.{TAXYR_FIELD}
-                """
-
+            # Construct the query to select the test results that were stored
+            # in Athena. We want to rehydrate a few fields like township_code
+            # and class without requiring that all tests select them, so
+            # check if those fields are missing and join against the correct
+            # source table in an attempt to rehydrate the missing columns.
+            test_results_select = "select test_results.*"
+            test_results_join = ""
+            # We need parid and taxyr at minimum in order to rehydrate any
+            # missing fields
+            if PARID_FIELD in fieldnames and TAXYR_FIELD in fieldnames:
+                if TOWNSHIP_FIELD not in fieldnames:
+                    test_results_select += f", leg.user1 AS {TOWNSHIP_FIELD}"
+                    test_results_join += f"""
+                        left join iasworld.legdat as leg
+                            on leg.{PARID_FIELD} = test_results.{PARID_FIELD}
+                            and leg.{TAXYR_FIELD} = test_results.{TAXYR_FIELD}
+                            and leg.cur = 'Y'
+                            and leg.deactivat is null
+                    """
+                if CLASS_FIELD not in fieldnames:
+                    if CARD_FIELD in fieldnames:
+                        # Figure out the right table to join to in order to
+                        # query the class
+                        card_join_table = "dweldat"
+                        if tablename in ["comdat", "land", "oby"]:
+                            card_join_table = tablename
+
+                        test_results_select += f", card.class AS {CLASS_FIELD}"
+                        test_results_join += f"""
+                            left join iasworld.{card_join_table} as card
+                                on card.{PARID_FIELD} =
+                                    test_results.{PARID_FIELD}
+                                and card.{TAXYR_FIELD} =
+                                    test_results.{TAXYR_FIELD}
+                                and card.{CARD_FIELD} =
+                                    test_results.{CARD_FIELD}
+                                and card.cur = 'Y'
+                                and card.deactivat is null
+                        """
+                    else:
+                        test_results_select += f", par.class AS {CLASS_FIELD}"
+                        test_results_join += f"""
+                            left join iasworld.pardat as par
+                                on par.{PARID_FIELD} =
+                                    test_results.{PARID_FIELD}
+                                and par.{TAXYR_FIELD} =
+                                    test_results.{TAXYR_FIELD}
+                                and par.cur = 'Y'
+                                and par.deactivat is null
+                        """
```

Comment:
I wonder if there is a cleaner way to do this?

Reply:
This is horrendous but also I love it. If it works it works. The only way I can think of that might be simpler is to create a "helper view" i.e. a view with the universe of township and class codes by table:

| PIN   | CARD | TAXYR | TOWNSHIP_CODE | TABLE   | CLASS |
|-------|------|-------|---------------|---------|-------|
| 12345 | 1    | 2024  | 10            | LAND    | 200   |
| 12345 | 1    | 2024  | 10            | DWELDAT | 211   |

Then filter the table and join it to get the town and class.

Reply:
Very smart! I like the idea of a helper table a lot but I don't want to hold up the PR for it. I left a comment to this effect in 77305a584265f770a3b4df6be50d5b1277c03fa3, we can pick it up next time we find ourselves needing to adjust this logic.